README.md
# Lambda Container Service
> Service template using container based lambda function

## Developing

Runs a custom dev webserver which hosts the lambda function which can be hosted on  
```shell
$ npm start
```

## Test

### Unit Tests

```shell
$ npm run test
```

### Integration Tests

#### Automated Integration Tests (WIP)
Runs service using `dev-server` and tests using Newman test collection.

```shell
$ npm run test:newman
```
#### Manual Test
Start the environment
```shell
$ npm run build
$ docker-compose build
$ docker-compose up
```
In another temrinal, run:

```shell
$ curl -XPOST "http://localhost:9000/2015-03-31/functions/function/invocations" -d '{}'
```

## Package @todo
- Check if container repository exists (Creates if not exists) - Should be in pipeline helpers
- npm run build
- npm run package
- create a release tag on Github Repo
- publish image based on version specified by semantic release.

## Deploy @todo

- Deploy a version to service's lambda by updating the image to release build version.
- And verify health check.

CHANGELOG.md
# [1.3.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.2.1...v1.3.0) (2021-11-22)


### Bug Fixes

* **newman:** fixed dev server express mapping ([9be5ee3](https://github.com/phenixcoder/lambda-container-service/commit/9be5ee3804eeec84316f36011d3c39e611d9e07c))


### Features

* **logger:** added improved logger ([2530e38](https://github.com/phenixcoder/lambda-container-service/commit/2530e3832e5ad53bf0b7e76bb26d6c574720f8a1))

# [1.3.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.2.1...v1.3.0) (2021-11-22)


### Bug Fixes

* **newman:** fixed dev server express mapping ([9be5ee3](https://github.com/phenixcoder/lambda-container-service/commit/9be5ee3804eeec84316f36011d3c39e611d9e07c))


### Features

* **logger:** added improved logger ([2530e38](https://github.com/phenixcoder/lambda-container-service/commit/2530e3832e5ad53bf0b7e76bb26d6c574720f8a1))

# [1.3.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.2.1...v1.3.0) (2021-11-22)


### Bug Fixes

* **newman:** fixed dev server express mapping ([9be5ee3](https://github.com/phenixcoder/lambda-container-service/commit/9be5ee3804eeec84316f36011d3c39e611d9e07c))


### Features

* **logger:** added improved logger ([2530e38](https://github.com/phenixcoder/lambda-container-service/commit/2530e3832e5ad53bf0b7e76bb26d6c574720f8a1))

# [1.3.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.2.1...v1.3.0) (2021-11-22)


### Bug Fixes

* **newman:** fixed dev server express mapping ([9be5ee3](https://github.com/phenixcoder/lambda-container-service/commit/9be5ee3804eeec84316f36011d3c39e611d9e07c))


### Features

* **logger:** added improved logger ([2530e38](https://github.com/phenixcoder/lambda-container-service/commit/2530e3832e5ad53bf0b7e76bb26d6c574720f8a1))

## [1.2.1](https://github.com/phenixcoder/lambda-container-service/compare/v1.2.0...v1.2.1) (2021-11-14)


### Bug Fixes

* **logger:** Removed file loggers ([f484141](https://github.com/phenixcoder/lambda-container-service/commit/f4841410d0c8f5e7aca3a9082809135f795f4e0e))

# [1.2.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.1.0...v1.2.0) (2021-11-10)


### Features

* **publish:** migrated to private ECR Repo ([77f4044](https://github.com/phenixcoder/lambda-container-service/commit/77f4044949dbdd375d4f51dd15bf0ab20f334fd1))

# [1.2.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.1.0...v1.2.0) (2021-11-10)


### Features

* **publish:** migrated to private ECR Repo ([77f4044](https://github.com/phenixcoder/lambda-container-service/commit/77f4044949dbdd375d4f51dd15bf0ab20f334fd1))

# [1.1.0](https://github.com/phenixcoder/lambda-container-service/compare/v1.0.0...v1.1.0) (2021-10-19)


### Bug Fixes

* **newman:** fixed newman test failing due to tsc build config ([13186d2](https://github.com/phenixcoder/lambda-container-service/commit/13186d28968ef5cb7240e52cff4c5ea99dc8ce91))


### Features

* **health-check:** add health check route to index handler ([bf2d6e2](https://github.com/phenixcoder/lambda-container-service/commit/bf2d6e2a81f0d3e82972d87aae6d613cb8be016e))

# 1.0.0 (2021-10-16)


### Features

* initial Commit ([dbde8fe](https://github.com/phenixcoder/lambda-container-service/commit/dbde8fefc96567b482a61682474a9a65679a74f0))


app/index.test.ts
import {
  APIGatewayEventRequestContext,
  APIGatewayProxyEventV2,
} from 'aws-lambda';
import { version, name } from '../package.json';
import { handler } from '.';

function mockContext(): APIGatewayEventRequestContext {
  return {
    accountId: 'string',
    apiId: 'string',
    authorizer: {
      jwt: {
        claims: {
          foo: 'string',
        },
        scopes: [],
      },
    },
    protocol: '',
    domainName: '',
    domainPrefix: 'string',
    requestId: '',
    routeKey: '',
    stage: '',
    httpMethod: '',
    identity: {
      accessKey: '',
      accountId: '',
      apiKey: '',
      apiKeyId: '',
      caller: '',
      clientCert: null,
      cognitoAuthenticationProvider: '',
      cognitoAuthenticationType: '',
      cognitoIdentityId: '',
      cognitoIdentityPoolId: '',
      principalOrgId: '',
      sourceIp: '0.0.0.0',
      user: '',
      userAgent: '',
      userArn: '',
    },
    path: '',
    requestTimeEpoch: 0,
    resourceId: '',
    resourcePath: '',
  };
}

function mockRequest(): APIGatewayProxyEventV2 {
  return {
    version: '',
    routeKey: '',
    rawPath: '',
    rawQueryString: '',
    headers: {},
    requestContext: {
      accountId: 'string',
      apiId: 'string',
      authorizer: {
        jwt: {
          claims: {
            foo: 'string',
          },
          scopes: [],
        },
      },
      domainName: '',
      domainPrefix: 'string',
      timeEpoch: 0,
      http: {
        method: 'string',
        path: 'string',
        protocol: 'string',
        sourceIp: 'string',
        userAgent: 'string',
      },
      requestId: '',
      routeKey: '',
      stage: '',
      time: '',
    },
    isBase64Encoded: false,
  };
}

describe('Index Handler', () => {
  it('Should return success response', async () => {
    const response = await handler(mockRequest(), mockContext());
    expect(response).toStrictEqual({
      body: JSON.stringify({
        hello: 'world',
        foo: 'bar',
      }),
      headers: {
        'content-type': 'text/json',
      },
      statusCode: 200,
    });
  });
  it('Should return help response when path is __health', async () => {
    const req = mockRequest();
    req.rawPath = '__health';
    const response = await handler(req, mockContext());
    expect(response).toStrictEqual({
      body: JSON.stringify({
        version,
        name,
      }),
      headers: {
        'content-type': 'text/json',
      },
      statusCode: 200,
    });
  });
});


src/index.test.ts
import {
  APIGatewayEventRequestContext,
  APIGatewayProxyEventV2,
} from 'aws-lambda';
import { version, name } from '../package.json';
import { handler } from '.';

function mockContext(): APIGatewayEventRequestContext {
  return {
    accountId: 'string',
    apiId: 'string',
    authorizer: {
      jwt: {
        claims: {
          foo: 'string',
        },
        scopes: [],
      },
    },
    protocol: '',
    domainName: '',
    domainPrefix: 'string',
    requestId: '',
    routeKey: '',
    stage: '',
    httpMethod: '',
    identity: {
      accessKey: '',
      accountId: '',
      apiKey: '',
      apiKeyId: '',
      caller: '',
      clientCert: null,
      cognitoAuthenticationProvider: '',
      cognitoAuthenticationType: '',
      cognitoIdentityId: '',
      cognitoIdentityPoolId: '',
      principalOrgId: '',
      sourceIp: '0.0.0.0',
      user: '',
      userAgent: '',
      userArn: '',
    },
    path: '',
    requestTimeEpoch: 0,
    resourceId: '',
    resourcePath: '',
  };
}

function mockRequest(): APIGatewayProxyEventV2 {
  return {
    version: '',
    routeKey: '',
    rawPath: '',
    rawQueryString: '',
    headers: {},
    requestContext: {
      accountId: 'string',
      apiId: 'string',
      authorizer: {
        jwt: {
          claims: {
            foo: 'string',
          },
          scopes: [],
        },
      },
      domainName: '',
      domainPrefix: 'string',
      timeEpoch: 0,
      http: {
        method: 'string',
        path: 'string',
        protocol: 'string',
        sourceIp: 'string',
        userAgent: 'string',
      },
      requestId: '',
      routeKey: '',
      stage: '',
      time: '',
    },
    isBase64Encoded: false,
  };
}

describe('Index Handler', () => {
  it('Should return success response', async () => {
    const response = await handler(mockRequest(), mockContext());
    expect(response).toStrictEqual({
      body: JSON.stringify({
        version,
        name,
        event: mockRequest(),
        context: mockContext(),
      }),
      headers: {
        'content-type': 'text/json',
      },
      statusCode: 200,
    });
  });
  it('Should return help response when path is __health', async () => {
    const req = mockRequest();
    req.rawPath = '__health';
    const response = await handler(req, mockContext());
    expect(response).toStrictEqual({
      body: JSON.stringify({
        version,
        name,
      }),
      headers: {
        'content-type': 'text/json',
      },
      statusCode: 200,
    });
  });
});


src/lib/logger.test.ts
import { level } from './logger';

describe('Logger', () => {
  it('level', () => {
    const testEnv = process.env.NODE_ENV;
    process.env.NODE_ENV = '';
    expect(level()).toBe('debug');
    process.env.NODE_ENV = 'production';
    expect(level()).toBe('warn');
    process.env.NODE_ENV = 'development';
    expect(level()).toBe('debug');
    process.env.NODE_ENV = testEnv;
  });
});


src/index.ts
import {
  APIGatewayEventRequestContext,
  APIGatewayProxyEventV2,
  APIGatewayProxyResultV2,
} from 'aws-lambda';
import Logger from './lib/logger';
import pkg from '../package.json';

export const handler = async (
  event: APIGatewayProxyEventV2,
  _context: APIGatewayEventRequestContext
): Promise<APIGatewayProxyResultV2> => {
  if (event.rawPath === '__health') {
    return {
      statusCode: 200,
      headers: {
        'content-type': 'text/json',
      },
      body: JSON.stringify({
        version: pkg.version,
        name: pkg.name,
      }),
    };
  }

  Logger.debug('This is a log message from handler', event);
  return {
    statusCode: 200,
    headers: {
      'content-type': 'text/json',
    },
    body: JSON.stringify({
      version: pkg.version,
      name: pkg.name,
      event,
      context: _context,
    }),
  };
};


src/lib/logger.ts
import winston from 'winston';

const levels = {
  error: 0,
  warn: 1,
  info: 2,
  http: 3,
  debug: 4,
};

export const level = () => {
  const env = process.env.NODE_ENV || 'development';
  const isDevelopment = env === 'development';
  return isDevelopment ? 'debug' : 'warn';
};

const colors = {
  error: 'red',
  warn: 'yellow',
  info: 'green',
  http: 'magenta',
  debug: 'white',
};

winston.addColors(colors);

const format = winston.format.combine(
  winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss:ms' }),
  winston.format.colorize({ all: true }),
  winston.format.printf(
    (info) => `${info.timestamp} ${info.level}: ${info.message}`
  )
);

const transports = [new winston.transports.Console()];

const Logger = winston.createLogger({
  level: level(),
  levels,
  format,
  transports,
});

export default Logger;


dev-server.js
const express = require('express')
const { existsSync } = require('fs')
const { handler } = require('./build/src/index')
const newman = require('newman')

const app = express();
const port = 9000;

let server = null;

app.post('/2015-03-31/functions/function/invocations', async (req, res) => {
  try {
    const body = req.body ? JSON.parse(req.body) : ExpressToAWSAPIGatewayProxyEventV2(req);
    handler(body, {}).then(result => {
      res.send(result);
    }).catch(error => console.error(error));
  } catch(error) {
    console.error(error);
  }
})

app.all('/api', async (req, res) => {
  handler(ExpressToAWSAPIGatewayProxyEventV2(req), {}).then(result => {
    res.send(result);
  }).catch(error => console.error(error));
});
app.all('/__health', async (req, res) => {
  handler(ExpressToAWSAPIGatewayProxyEventV2(req), {}).then(result => {
    res.send(result);
  }).catch(error => console.error(error));
});

server = app.listen(port, () => {
  console.log(`
  Example app listening at: 
    POST http://localhost:${port}/2015-03-31/functions/function/invocations
  `);

  if (process.argv[2]) {
    const testFilename = process.argv[3] || 'service-collection.postman_collection.json'

    if (!existsSync(`./${testFilename}`)) {
    console.log('Running Newman tests');
      console.error(`Error: test collection not found
        ./${testFilename}
      
        Skipping tests. Running dev server only.
      `);

      return;
    }
    console.log('Test collection:', testFilename);

    newman.run({
      collection: require(`./${testFilename}`),
      reporters: 'cli',
    }, 
    (err, summary) => {
      if (err) {
        console.log(err);
        process.exit(1)
      }
      process.exit();
    })
  }
})

function ExpressToAWSAPIGatewayProxyEventV2(request) {
  return {
    "resource": request.path,
    "path": request.path,
    "httpMethod": request.method,
    "headers": {
      "header1": "value1",
      "header2": "value2"
    },
    "multiValueHeaders": {
      "header1": [
        "value1"
      ],
      "header2": [
        "value1",
        "value2"
      ]
    },
    "queryStringParameters": {
      "parameter1": "value1",
      "parameter2": "value"
    },
    "multiValueQueryStringParameters": {
      "parameter1": [
        "value1",
        "value2"
      ],
      "parameter2": [
        "value"
      ]
    },
    "requestContext": {
      "accountId": "123456789012",
      "apiId": "id",
      "authorizer": {
        "claims": null,
        "scopes": null
      },
      "domainName": "id.execute-api.us-east-1.amazonaws.com",
      "domainPrefix": "id",
      "extendedRequestId": "request-id",
      "httpMethod": "GET",
      "identity": {
        "accessKey": null,
        "accountId": null,
        "caller": null,
        "cognitoAuthenticationProvider": null,
        "cognitoAuthenticationType": null,
        "cognitoIdentityId": null,
        "cognitoIdentityPoolId": null,
        "principalOrgId": null,
        "sourceIp": "IP",
        "user": null,
        "userAgent": "user-agent",
        "userArn": null,
        "clientCert": {
          "clientCertPem": "CERT_CONTENT",
          "subjectDN": "www.example.com",
          "issuerDN": "Example issuer",
          "serialNumber": "a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1:a1",
          "validity": {
            "notBefore": "May 28 12:30:02 2019 GMT",
            "notAfter": "Aug  5 09:36:04 2021 GMT"
          }
        }
      },
      "path": "/my/path",
      "protocol": "HTTP/1.1",
      "requestId": "id=",
      "requestTime": "04/Mar/2020:19:15:17 +0000",
      "requestTimeEpoch": 1583349317135,
      "resourceId": null,
      "resourcePath": "/my/path",
      "stage": "$default"
    },
    "pathParameters": null,
    "stageVariables": null,
    "body": "Hello from Lambda!",
    "isBase64Encoded": false
  }
}